src/main/java/uk/gov/companieshouse/api/accounts/util/logging/ControllerLoggingAspect.java

```diff
@@ -0,0 +1,87 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+import org.aspectj.lang.JoinPoint;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.AfterReturning;
+import org.aspectj.lang.annotation.AfterThrowing;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.annotation.Before;
+import org.aspectj.lang.annotation.Pointcut;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+
+@Aspect
+@Component
+public class ControllerLoggingAspect {
+
+  private static final String REQUEST_ID = "X-Request-Id";
+
+  private static final String ERIC_IDENTITY = "Eric-Identity";
+
+  @Pointcut("execution(* uk.gov.companieshouse.api.accounts.controller.*.*(..))")
+  private void controllerMethods() {
+    //Defines the point cut for controller methods, the method body kept empty on purpose.
+  }
+
+  @Around("controllerMethods()")
+  public void logPerformanceStats(ProceedingJoinPoint joinPoint) throws Throwable {
+    long startTime = System.currentTimeMillis();
+    //Do not change, the target method will not progress without this.
+    Object result = joinPoint.proceed();
+    long responseTime = System.currentTimeMillis() - startTime;
+
+    LogBuilder logBuilder = getLogBuilder(joinPoint);
+    logBuilder.logStartOfRequestProcessing(String.format("Start Request %s",
+        methodName(joinPoint)));
+
+    logBuilder.logEndOfRequestProcessing(String.format("End of Request %s Successfully Completed",
+        methodName(joinPoint)), statusCode(result), responseTime);
+  }
+
+  private String methodName(JoinPoint joinPoint) {
+    return joinPoint.getSignature().getName();
+  }
+
+  private String requestId(HttpServletRequest request) {
+    return request.getHeader(REQUEST_ID);
+  }
+
+  private String userId(HttpServletRequest request) {
+    return request.getHeader(ERIC_IDENTITY);
+  }
+
+  private LogBuilder getLogBuilder(JoinPoint joinPoint) {
+    HttpServletRequest request = getRequest(joinPoint);
+    String accountId = "";
+    return new LogBuilderImpl(requestId(request), userId(request),
+        transactionId(joinPoint), accountId);
+  }
+
+  /**
+   * Helper to get the HttpServletRequest so that we can get the requestID and the userID.
+   */
+  private HttpServletRequest getRequest(JoinPoint joinPoint) {
+    return (HttpServletRequest) joinPoint.getArgs()[0];
+  }
+
+  private String transactionId(JoinPoint joinPoint) {
+    return joinPoint.getArgs()[1].toString();
+  }
+
+  private int statusCode(Object result) {
+    assert(result instanceof ResponseEntity);
+    return ((ResponseEntity) result).getStatusCode().value();
```

Comment:
What happens when an AssertionError is thrown? The globalExceptionHandler is catching Exceptions but not Errors. Will this be correctly handled and clearly logged?

From [Oracle docs](https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html#usage):

> Another problem with using assertions for argument checking is that erroneous arguments should result in an appropriate runtime exception (such as IllegalArgumentException, IndexOutOfBoundsException, or NullPointerException). An assertion failure will not throw an appropriate exception.

Reply:
will change it